### PR TITLE
Added support for SparkSession

### PR DESCRIPTION
**Pull request by rozza**
Deprecated public methods using SQLContext. In Spark 2.0 SQLContext
is a wrapper of SparkSession and is the preferred class.

SPARK-47

### CODE REVIEW

File: src/main/scala/com/mongodb/spark/MongoSpark.scala [hunk 1]
```diff
@@ -390,6 +394,49 @@ object MongoSpark {
     save[D](javaRDD.rdd, writeConfig)
   }
 
+  // Deprecated APIs
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/MongoSpark.scala]: Cool, looks like you're preserving binary compatibility.

File: src/main/scala/com/mongodb/spark/MongoSpark.scala [hunk 2]
```diff
@@ -399,10 +446,10 @@ object MongoSpark {
  *
  * @since 1.0
  */
-case class MongoSpark(sqlContext: SQLContext, connector: MongoConnector, readConfig: ReadConfig, pipeline: Seq[BsonDocument]) {
+case class MongoSpark(sparkSession: SparkSession, connector: MongoConnector, readConfig: ReadConfig, pipeline: Seq[BsonDocument]) {
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/MongoSpark.scala]: Not preserving it here though?

File: src/main/scala/com/mongodb/spark/sql/package.scala [hunk 3]
```diff
@@ -31,7 +31,15 @@ package object sql {
    * @param sqlContext the current SQLContext
    * @return the MongoDB based SQLContext
    */
-  implicit def toSparkSQLContextFunctions(sqlContext: SQLContext): SparkSQLContextFunctions = SparkSQLContextFunctions(sqlContext)
+  implicit def toSparkSessionFunctions(sqlContext: SQLContext): SparkSessionFunctions = toSparkSessionFunctions(sqlContext.sparkSession)
```
Comment (jyemin) [inline src/main/scala/com/mongodb/spark/sql/package.scala]: Can this be deprecated?

**Comment by jyemin** [comment]
Looks good.  A couple of comments/questions:

1. Looks like there are some methods that take/return a SQLContext that have been deprecated along with new methods that take/return a SparkSession, and some that have just been changed from SQLContext to SparkSession, which breaks compatibility.  I'm not clear on your thinking here.  Can you explain what you are trying to achieve?

2. Are any of the deprecated methods no longer tested?

File: src/main/scala/com/mongodb/spark/sql/package.scala (hunk 3 above)
Comment (rozza) [inline src/main/scala/com/mongodb/spark/sql/package.scala]: Its an implicit method, so I didn't but I suppose we can.

**Comment by rozza** [comment]
1. We are breaking compatibility with the 2.0 connector - as Spark itself causes breaks.
    Public methods not covered by the @DeveloperApi annotation have been deprecated. Internal or 
   development api's have been switched to use the new methods.

2. All tests have been updated to use `SparkSession` instead.

**Comment by jyemin** [comment]
lgtm then

**Comment by rozza** [comment]
This was rebased and merged.

**Pull request closed without merging**